### ISSUE TITLE
Linear fixed-point solver in backward

### ISSUE THREAD

**Issue by mcbal** [locuslab/mdeq#8]
Hi @jerrybai1995,

Thanks a lot for sharing the `deq` and `mdeq` repositories!

After reading the [Deep Implicit Layers tutorial](https://implicit-layers-tutorial.org/deep_equilibrium_models/) I had two small questions regarding the implementation of the linear fixed-point equation solving happening in the backward pass.

In both [MDEQ](https://github.com/locuslab/mdeq/blob/eb3d85fa01404fdaaa4913fc220653ffee354078/lib/modules/deq2d.py#L151-L155) and [DEQ](https://github.com/locuslab/deq/blob/dd3b8dddf18a0f5353b05a71cc7df278f22339ec/DEQModel/modules/deq.py#L109-L113) a function `g(x)` is defined to solve the linear fixed-point equations. For MDEQ it looks like this:

https://github.com/locuslab/mdeq/blob/eb3d85fa01404fdaaa4913fc220653ffee354078/lib/modules/deq2d.py#L151-L155

Now `broyden` solves for `g(x)=0` (like done in the forward pass at equilibrium). But to agree with the fixed-point equations, wouldn't there need to be an additional term `-x` added to `res`? It looks like it's now solving for a vector that makes the VJP equal to `-grad`. I got confused because `g(x)` looks very similar to the `lambda`-function defined in the [NeurIPS tutorial using an Anderson fixed-point solver](http://implicit-layers-tutorial.org/deep_equilibrium_models/) but that solver solves for `f(x)=x` instead of `g(x)=f(x)-x=0`.

My second question is about the lack of `.clone().detach()` in the snippet above. Is there a reason for the `.clone().detach()` appearing the DEQ implementation in the respective definition of `g(x)` but not in the MDEQ implementation?

Thank you!

**Comment by shaojieb** [comment]
Hi @mcbal ,

Thanks for the questions. 

1. No. In the backward pass we want to solve for Jx+grad=x. You can either consider it as a fixed point solving process f(x)=x where f(x)=Jx+grad, or a root solving process g(x)=0 where g(x)=(J-I)x+grad. In this case, we consider the latter, and with the Jacobian defined by https://github.com/locuslab/mdeq/blob/eb3d85fa01404fdaaa4913fc220653ffee354078/lib/modules/deq2d.py#L149, we are simply computing (J-I)x+grad in L153 (of your screenshot above). There is no need to `-x` if we use Broyden.

2. There is no difference. I was only adding that in the DEQ repo for safety purposes. In fact, following the NeurIPS tutorial, we have updated the code to use backward hook, rather than DummyBackward, to do the backward pass. This is in the `beta` branch of the DEQ repo: https://github.com/locuslab/deq/tree/beta. The backward pass for a DEQ-Transformer is as simple as https://github.com/locuslab/deq/blob/beta/DEQModel/models/transformers/deq_transformer.py#L346. Feel free to check it out!

Let me know if this clarifies the questions you have :-)

**Comment by mcbal** [comment]
Hi @jerrybai1995 ,

Thanks a lot for the quick and detailed reply!

1. Thanks for the explicit walk-through, it cleared up my confusion.
2. Wonderful.

All clear, so closing this :)
